### PR TITLE
Druid support via SQLAlchemy

### PR DESCRIPTION
**Pull request by betodealmeida**
I recently created a module called [`druiddb`](https://github.com/betodealmeida/druid-dbapi) (merged into `pydruid` this week) that provides a SQLAlchemy dialect for Druid. This allows Superset to talk to Druid using its standard SQLAlchemy connector, instead of the custom one.

One problem with this approach is that Druid does not support joins, and some queries (timeseries with limit) perform an inner join to get the top overall groups. In order to handle Druid correctly I added a new attribute to engine specs called `inner_joins`, defaulting to true.

If this attribute is false, instead of building the inner join we run a "prequery", fetching the top groups similar to how the Druid connector works. The values are then used as an extra filter in the main query. Eg, this is how a query **with** join works (from the `birth_names` dataset):

```sql
SELECT name AS name,
       ds AS __timestamp,
       SUM(birth_names.num) AS sum__num
FROM birth_names
JOIN
  (SELECT name AS name__,
          SUM(birth_names.num) AS mme_inner__
   FROM birth_names
   WHERE ds >= '1918-01-05 00:00:00.000000'
     AND ds <= '2018-01-05 11:59:32.000000'
   GROUP BY name
   ORDER BY mme_inner__ DESC
   LIMIT 5
   OFFSET 0) AS anon_1 ON name = name__
WHERE ds >= '1918-01-05 00:00:00.000000'
  AND ds <= '2018-01-05 11:59:32.000000'
GROUP BY name,
         ds
ORDER BY sum__num DESC
LIMIT 50000
OFFSET 0;
```

And here how it looks when inner joins are **not supported**:

```sql
SELECT name AS name,
       SUM(birth_names.num) AS sum__num
FROM birth_names
WHERE ds >= '1918-01-05 00:00:00.000000'
  AND ds <= '2018-01-05 12:00:29.000000'
GROUP BY name
ORDER BY SUM(birth_names.num) DESC
LIMIT 5
OFFSET 0;

SELECT name AS name,
       ds AS __timestamp,
       SUM(birth_names.num) AS sum__num
FROM birth_names
WHERE ds >= '1918-01-05 00:00:00.000000'
  AND ds <= '2018-01-05 12:00:29.000000'
  AND (name = 'Michael'
       OR name = 'Christopher'
       OR name = 'David'
       OR name = 'James'
       OR name = 'John')
GROUP BY name,
         ds
ORDER BY sum__num DESC
LIMIT 50000
OFFSET 0;
```

Both queries are shown when clicking "View Query", instead of only the last one. See the screenshot:

screen shot 2018-01-05 at 12 24 41 pm

In order to do that, I added two new arguments to the query object:

1. `prequeries` is a list that stores prequeries;
2. `is_prequery` is a boolean indicating if a given query is the final one, or a prequery.

When a main query runs a prequery, it will append it to `prequeries`. The functions `query` and `get_query_string_response` then take care of combining prequeries with the main query, so it can be displayed to the user correctly.

### CODE REVIEW

**Comment by betodealmeida** [comment]
I'll fix the unit tests.

File: superset/connectors/sqla/models.py [hunk 1]
```diff
@@ -528,37 +533,73 @@ def get_sqla_query(  # sqla
...
+                    'orderby': orderby,
+                    'extras': extras,
+                    'columns': columns,
+                    'form_data': form_data,
+                    'order_desc': True,
+                }
+                result = self.query(subquery_obj)
+                dimensions = [c for c in result.df.columns if c not in metrics]
+                top_groups = self._get_top_groups(result.df, dimensions)
+                qry = qry.where(top_groups)
```
Comment (mistercrunch) [inline superset/connectors/sqla/models.py]: If `where` is called multiple times, does SQLAlchemy goes for a logical `AND`? Couldn't find the documentation for that method quickly...

Comment (mistercrunch) [reply superset/connectors/sqla/models.py]: Found the answer, logical AND is applied

**Comment by mistercrunch** [comment]
LGTM

File: superset/connectors/sqla/models.py (hunk 1 above)
Comment (betodealmeida) [reply superset/connectors/sqla/models.py]: For [reference](http://docs.sqlalchemy.org/en/latest/core/selectable.html#sqlalchemy.sql.expression.Select.where):

> return a new select() construct with the given expression added to its WHERE clause, **joined to the existing clause via AND, if any**. (emphasis mine)

**Pull request merged**